### PR TITLE
Add footstep sounds for liquids

### PR DESCRIPTION
**Pull request by MirceaKitsune**
As promised with my underwater view bobbing patch, I added some footstep sounds for liquids which will be heard when swimming. The sounds were recorded by me and licensed WTFPL (or whatever works best).

### CODE REVIEW

**Comment by Wuzzy2** [comment]
Ugh, sorry, but the sounds are awful for me.
It sounds like if you are swimming in a bathtub made out of rusty iron rather than in open water.

Also: You used the same water sound for lava. o_O

**Comment by MirceaKitsune** [comment]
@Wuzzy2 I think the sounds should be acceptable, although I'm aware they're not the best. If anyone can find better ones, please send them to me and I'll gladly include them instead.

I wasn't sure if there's a point in making separate sounds for lava, nor how to achieve them at all. I'm not even sure how swimming in lava would sound like to be honest. Once again, if anyone finds some mention them to me and I will include them.

**Comment by BlockMen** [comment]
In general yes, but the sounds are horrible.
Also it should not be used by lava. Better no "lava swimming" sound than a water like.

**Comment by MirceaKitsune** [comment]
@BlockMen I can't see how much better ones are possible, although somewhat better sounds would be doable. I looked on http://www.pdsounds.org/ but didn't find anything good (their search system is also very broken). Anyone else have any ideas?

I think it's more silly to have water sounds but no lava sounds personally... even if it's less likely for anyone to go swimming in lava. What would indeed work better are some swamp water sounds, which could also be used for actual swamp water (custom biomes).

**Comment by MirceaKitsune** [comment]
Okay, I found a few good ones on OpenGameArt. Will choose what sounds best and update.

Water:
http://opengameart.org/content/skippy-fish-water-sound-collection
http://opengameart.org/content/water-splashes

Lava:
http://opengameart.org/content/lava-splash
http://opengameart.org/content/acid-burn-sounds

**Comment by MirceaKitsune** [comment]
Done. I got rid of my ugly sounds and included much nicer ones from OGA (licensed CC-BY 3.0). Go ahead an give them another listen :)

**Comment by Wuzzy2** [comment]
Ah, water is better now. But the 2 lava sounds sound almost identical to me.

**Comment by MirceaKitsune** [comment]
There's still a problem with the patch: I'm trying to make the sounds work for buckets as well, so the sound is also heard when using a bucket to drop or take liquids. I experimentally replaced add_node with place_node and remove_node with dig_note in the bucket mod, but for some reason this doesn't cause the place / dig sounds to play.

Shouldn't creating a block with place_node cause the node's on_place sound to be heard, whereas add_node shouldn't? The description does say that place_node is "as if a player has put the node there". Same with dig_node over remove_node.

If not, is there another way (that's flexible and acceptable) to solve the issue? I'm trying the above to avoid coding the sounds directly into the buckets, which is an easy solution but less flexible. I'd rather trigger the source node's place / dig sound, but is this possible with buckets?

**Comment by thatgraemeguy** [comment]
Shouldn't the lava sound be a bloodcurdling scream? I imagine that'd be the only sound I'd hear if I were strolling through lava.

**Comment by MirceaKitsune** [comment]
@thatgraemeguy Using voices for sound effects is likely not acceptable. Especially because they'd have to be gender neutral, and even species neutral considering mods that let you play as animals or monsters (like my Creatures mod). Other mods can add their own hurt sounds of course.

That said, these are the best sounds I could possibly find, and I think the current lava sounds are really good. Especially since lava is dense and you swim through it very slowly, causing the sound to play rarely and usually just when you swim near the surface or fall in.

**Comment by C1ffisme** [comment]
@thatgraemeguy I think a bloodcurdling scream would scare younger players away and might scare parents from letting their kids play.

**Comment by tenplus1** [comment]
This feature has been done already by using the Ambiance mod... https://github.com/tenplus1/ambience

**Comment by MirceaKitsune** [comment]
@tenplus1 The ambiance mod is likely not going to be part of minetest_game, and like all mods not every server will use it. This is an addition for default MTG, aiming to introduce a needed feature that's long been missing.

Also, I remember the Ambience mod loops a single swimming sound while you're in water, whereas this uses node footsteps which is more correct and reliable. They didn't work until a few weeks ago however, since footsteps for liquids were broken in the engine and required a patch to enable... this is in a sense the second part of the same improvement.

**Comment by BlockMen** [comment]
The watersound is better, but still not good. IMO not good enough to be merged.
Also, there is still a swimming sound for lava. This is just wrong.

**Comment by MirceaKitsune** [comment]
@BlockMen Better sounds for water are probably impossible to find. Those are from OGA, and really the best I've heard.

The swimming sound for lava is usually triggered only when you fall in. Lava is very dense and your swimming speed in it very slow, so it only plays on occasion. I think it's very realistic this way. If anyone wants to try it out, disable damage find a deep cave and go for a lava swim :)

**Comment by BlockMen** [comment]
@MirceaKitsune I actually did that (swimming in lava) to test it. And no, it did not convince me.

I doubt there are no better sounds. But these sounds would not improve the game, they would strengthen the (possible) impression of a cheap clone.

**Comment by MirceaKitsune** [comment]
Anything new on this? I understand water is good, but some people are still not sure about the lava. Can more developers mention their opinion please?

To easily test when and how the lava sounds are played: Disable damage and start any world. Go into noclip mode and fly underground until you find a cave with a large lava pool. Disable fly / noclip mode then jump in and swim around.

You'll only hear the sound once when you enter the lava, or while you swim and jump near the surface. Movement through it is so slow the footsteps are almost never triggered... which I think is great!

**Comment by MirceaKitsune** [comment]
Any more feedback on this, please? I understand the only remaining controversy is with the lava sounds... are they acceptable in the end, or should I remove them?

**Comment by Megaf** [comment]
Sorry, in my personal option I rather not have sound when moving on water. Perhaps if there is an option to enable/disable it would be a welcome feature.

**Comment by paramat** [comment]
@MirceaKitsune I feel there's no need for lava swmmiing sounds because the almost instant death means the extra code is not justified.

**Comment by HybridDog** [comment]
paramat, if damage is disabled you can swim there

**Comment by paramat** [comment]
I guess it might be good for when a plyer has lava resistence.
Lava sounds seem ok.
Water sounds are a little irritating every step, how about making them more random with a 67% or 50% chance?

**Comment by MirceaKitsune** [comment]
@paramat Making them random by including an empty sound sounds like a very ugly hack, which other developers would likely not agree to either. I mean some didn't agree with adding menu music because a simple song file was considered too large, it would be weird if this was considered okay instead. The only way to reduce footstep sound frequency is to increase the liquid viscosity so movement is slower, which is likely not desired.

These are changes I probably won't make. But I'm leaving this pull open since... in all honesty it's something simple and seems really good to me, so maybe others can come to an agreement or make the necessary changes.

**Comment by kilbith** [comment]
@paramat I like your sounds in your `trail` mod. I'd support that if you open a PR.

**Comment by paramat** [comment]
Do the changes to bucket code here need updating or removing? Do they actually work now?

File: mods/bucket/init.lua [hunk 1]
```diff
@@ -123,6 +123,7 @@ minetest.register_craftitem("bucket:bucket_empty", {
 				return
 			end
 
+			minetest.dig_node(pointed_thing.under)
 			minetest.add_node(pointed_thing.under, {name="air"})
```
Comment (HybridDog) [inline mods/bucket/init.lua]: minetest.dig_node already removes the node, it's not necessary to set "air", is it?

**Comment by 0-afflatus** [comment]
What water really needs is the ability to emit ambient sounds, but that's an engine issue.
Lots of sploshing noises will get annoying very fast, but sounds for emptying / filling buckets could be good.

**Comment by kilbith** [comment]
I still want the underwater's sounds in paramat's `trail` mod.

**Comment by paramat** [comment]
Okay i'll submit a PR to see what people think.

File: mods/bucket/init.lua [hunk 2]
```diff
@@ -75,7 +75,7 @@ function bucket.register_liquid(source, flowing, itemname, inventory_image, name
 							"place "..source) then
 						return
 					end
-					minetest.add_node(pos, {name=source})
+					minetest.place_node(pos, {name=source})
```
Comment (kaeza) [inline mods/bucket/init.lua]: I think this is out of scope for this commit. Maybe split?

File: mods/bucket/init.lua [hunk 3]
```diff
@@ -123,6 +123,7 @@ minetest.register_craftitem("bucket:bucket_empty", {
 				return
 			end
 
+			minetest.dig_node(pointed_thing.under)
```
Comment (kaeza) [inline mods/bucket/init.lua]: Same as above.

**Comment by paramat** [comment]
:-1:

File: mods/bucket/init.lua (hunk 2 above)
Comment (paramat) [inline mods/bucket/init.lua]: Agreed.

**Comment by sofar** [comment]
http://www.freesound.org/search/?q=underwater&f=license%3AAttribution&s=score+desc&advanced=0&g=1

sooo many good sounds, ready to use.

**Comment by paramat** [comment]
Closing for reasons given above and author refuses to improve the PR.
